### PR TITLE
Reduce Kernel Optimization

### PR DESCRIPTION
**Pull request by iK1D**
Our reduce kernel will check the reduce axes, if all the axes are contiguous, we will use our fast reduce kernel instead of the CUDNN implementation, which is much faster.

For some models exported from PyTorch with Torch.Tensor.repeat on first/last dim, we will use Tile Op for that, and the TileGrad is to reshape the GO to a shape combining input's shape and repeats, and then call ReduceSum on all even dims (start from 0). For example, if it's tensor's shape is [3,4,5] and call tensor.repeat(2,1,1), the ONNX gradient graph will reshape the GO to [2,3,1,4,1,5] and then call ReduceSum on axes [0,2,4], which is not contiguous. But actually we can remove all dims with value 1 from the shape as well as the corresponding axes from the attribute axes. In above example, ReduceSum to [2,3,1,4,1,5] on axes [0,2,4] is same as ReduceSum to [2,3,4,5] on axes [0]. With this change, we can use our fast reduce implementation.

This PR is to adjust the input shape and axes before kernel calculation.

With this fix, for DEBERTA model, it runs ~8% faster (from 3.17it/s to 3.41it/s for a certain config and env).

### CODE REVIEW

File: orttraining/orttraining/test/training_ops/cuda/reduce_sum_test.cc [hunk 1]
```diff
@@ -100,5 +100,29 @@ TEST(CudaKernelTest, ReduceSum_LargeTensorTrailingAxes) {
   TestReduceSum(X_dims, Y_dims, axes, keepdims);
 }
 
+TEST(CudaKernelTest, ReduceSum_OneDimsOptimization) {
+  std::vector<int64_t> X_dims{2, 3, 1, 4, 1, 5};
+  std::vector<int64_t> Y_dims{3, 4, 5};
+  std::vector<int64_t> axes{0, 2, 4};
+  bool keepdims = false;
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/training_ops/cuda/reduce_sum_test.cc:107]: just noticed there is no keepdims=true case in this test file...

File: onnxruntime/core/providers/cuda/reduction/reduction_functions.cc [hunk 2]
```diff
@@ -105,8 +105,42 @@ ApplicableMatrixReduction get_applicable_matrix_reduction(
...
+    if (dims[i] != 1) {
+      new_dims.emplace_back(dims[i]);
+      if (original_axes_set.find(gsl::narrow<int64_t>(i)) != original_axes_set.end()) {
+        new_axes.emplace_back(new_axis);
+      }
+      new_axis++;
+    }
+  }
+
+  // Empty axes means reduce all dimensions, which has different meaning,
```
Comment (SherlockNoMad) [inline onnxruntime/core/providers/cuda/reduction/reduction_functions.cc:130]: The default is to reduce over all the dimensions of the input tensor if 'noop_with_empty_axes' is false, else act as an Identity op when 'noop_with_empty_axes' is true.

I assume we have handled noop_with_empty_axes=true && axes.empty() case earlier in the code? If that's the case, maybe assert here?

File: onnxruntime/core/providers/cuda/reduction/reduction_functions.cc [hunk 3]
```diff
@@ -105,8 +105,42 @@ ApplicableMatrixReduction get_applicable_matrix_reduction(
...
+  // so add a new dim to the end if all original axes are on dims with value 1.
+  if (!original_axes.empty() && new_axes.empty()) {
```
Comment (SherlockNoMad) [inline onnxruntime/core/providers/cuda/reduction/reduction_functions.cc:132]: this case should translate into no-op... do we handle such case?

File: onnxruntime/core/providers/cuda/reduction/reduction_functions.cc [hunk 4]
```diff
@@ -105,8 +105,42 @@ ApplicableMatrixReduction get_applicable_matrix_reduction(
...
+    new_dims.emplace_back(1);
+    new_axes.emplace_back(new_axis);
+  }
+
+  // If all dims are value 1, make sure it's not empty by adding a new dim.
```
Comment (SherlockNoMad) [inline onnxruntime/core/providers/cuda/reduction/reduction_functions.cc:137]: this should also result in an no-op, also need a memcpy from input to output for this case.

**Review by SherlockNoMad — DISMISSED** [review]
LGTM

**Review by iK1D — COMMENTED** [review]
comment

File: onnxruntime/core/providers/cuda/reduction/reduction_functions.cc (hunk 3 above)
Comment (iK1D) [reply onnxruntime/core/providers/cuda/reduction/reduction_functions.cc:132]: By adding dummy dim and axis to the end, the compute will go on. For example, if shape=[2,1,1], and axes=[1,2], the new shape=[2,1] and new axes=[1], the compute will work as normal.

File: onnxruntime/core/providers/cuda/reduction/reduction_functions.cc (hunk 4 above)
Comment (iK1D) [reply onnxruntime/core/providers/cuda/reduction/reduction_functions.cc:137]: Same as above with different shapes: shape=[1,1], and axes=[], the new shape=[1] and new axes=[], the compute will work as normal.

**Comment by SherlockNoMad** [comment]
Sync offline, benefit of doing memcpy for elementwise copy is probably small. 
Good to merge after adding some UT case.

**Review by SherlockNoMad — APPROVED** [review]

**Pull request merged**